libshpool/src/config_watcher.rs

```diff
@@ -0,0 +1,676 @@
+                }
+            })
+            .context("create config reload thread")?;
+
+        Ok(Self {
+            tx: req_tx,
+            worker,
+            #[cfg(test)]
+            debug_rx,
+        })
+    }
+
+    /// Adds a watch on `path`.
+    ///
+    /// # Errors
+    /// Returns error if the underlying thread is gone, e.g. the worker thread encountered fatal
+    /// error and stopped its event loop.
+    pub fn watch(&self, path: impl AsRef<Path>) -> anyhow::Result<()> {
+        self.tx.send(path.as_ref().to_owned()).context("send AddWatch to ConfigWatcherInnfer")?;
+        Ok(())
+    }
+
+    /// Stops watching, shutting down the worker thread.
+    #[allow(dead_code)]
+    pub fn stop(self) {
+        mem::drop(self.tx);
+        self.worker.join().unwrap();
+    }
+
+    /// Worker is idle and ready for the next event. Debug/test only.
+    #[cfg(test)]
+    fn worker_ready(&self) {
+        self.debug_rx.recv().unwrap();
+        debug!("worker ready");
+    }
+}
+
+struct ConfigWatcherInner<Handler> {
+    // time to wait before actual reloading
+    reload_debounce: Duration,
+    // deadline to do a reload
+    reload_deadline: Option<Instant>,
+
+    handler: Handler,
+
+    watcher: INotifyWatcher,
+    // receiving notify events
+    notify_rx: Receiver<Result<notify::Event, notify::Error>>,
+    // receiving watch requests
+    req_rx: Receiver<PathBuf>,
+    // from target_path to (watched_path, immediate_child_path)
+    paths: HashMap<PathBuf, (PathBuf, PathBuf)>,
+
+    // for sending out debug info
+    #[cfg(test)]
+    debug_tx: Sender<()>,
+}
+
+// Outcomes of selecting channels in the worker thread
+enum Outcome {
+    // A notify event occurred
+    Event(notify::Result<notify::Event>),
+    // A control command from outside
+    AddWatch(PathBuf),
+    // Timeout on notify event, trigger reload
+    Timeout,
+    // Any channel was disconnected
+    Disconnected,
+}
+
+impl From<Result<notify::Result<notify::Event>, RecvError>> for Outcome {
+    fn from(value: Result<notify::Result<notify::Event>, RecvError>) -> Self {
+        match value {
+            Ok(v) => Self::Event(v),
+            Err(RecvError) => Self::Disconnected,
+        }
+    }
+}
+
+impl From<Result<PathBuf, RecvError>> for Outcome {
+    fn from(value: Result<PathBuf, RecvError>) -> Self {
+        match value {
+            Ok(v) => Self::AddWatch(v),
+            Err(RecvError) => Self::Disconnected,
+        }
+    }
+}
+
+impl<Handler> ConfigWatcherInner<Handler> {
+    // get next event to work on
+    fn select(&self) -> Outcome {
+        debug!("now {:?} select with ddl {:?}", Instant::now(), &self.reload_deadline);
+
+        // only impose a deadline if there is pending reload
+        let timeout = self
+            .reload_deadline
+            .map(crossbeam_channel::at)
+            .unwrap_or_else(crossbeam_channel::never);
+        // first try poll
+        if let Ok(res) = self.notify_rx.try_recv() {
```

Comment:
What is the purpose of doing these try_recvs early instead of going directly into the blocking wait? The blocking wait should return immediatly if there is data available. If it is just for the debug_tx stuff, you could wrap the whole thing in a block and put your `#[cfg(test)]` annotation on the block.

Reply:
Indeed, let me wrap the whole thing in a `#[cfg(test)]` block.